File: app/src/main/java/ght/app/datalogger/ChartFragment.java
package ght.app.datalogger;


import android.annotation.SuppressLint;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.anychart.AnyChart;
import com.anychart.AnyChartView;
import com.anychart.chart.common.dataentry.DataEntry;
import com.anychart.chart.common.dataentry.ValueDataEntry;
import com.anychart.charts.Cartesian;
import com.anychart.charts.Pie;
import com.anychart.core.cartesian.series.Line;
import com.anychart.data.Mapping;
import com.anychart.data.Set;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

import ght.app.datalogger.databinding.FragmentChartBinding;


public class ChartFragment extends Fragment {

    private FragmentChartBinding binding;

    @Override
    public void onCreate(@Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

    }

    @SuppressLint("ResourceType")
    @Nullable
    @org.jetbrains.annotations.Nullable
    @Override
    public View onCreateView(@NonNull @NotNull LayoutInflater inflater,
                             @Nullable @org.jetbrains.annotations.Nullable ViewGroup container,
                             @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {

        binding = FragmentChartBinding.inflate(inflater, container, false);

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view,
                              @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        //TODO: Somehow the charts are not shown in a virtual device.
        //      Test on real device
        createTestChart();

    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    private void createChart() {
        Cartesian chart = AnyChart.line();

        chart.animation();
        chart.title("Test trendview");
        chart.yAxis(0).title("Sensorwert");
        chart.xAxis(0).labels().padding(5d, 5d, 5d, 5d);

        List<DataEntry> data = new ArrayList<>();
        data.add(new ValueDataEntry("10-10-2020 10:10", 20));
        data.add(new ValueDataEntry("10-10-2020 10:11", 21));
        data.add(new ValueDataEntry("10-10-2020 10:12", 25));
        data.add(new ValueDataEntry("10-10-2020 10:13", 15));

        Set set = Set.instantiate();
        set.data(data);

        Mapping seriesMapping = set.mapAs("{x: 'x', value: 'value'");

        Line series = chart.line(seriesMapping);

        chart.legend().enabled();
        chart.legend().fontSize(13d);
        chart.legend().padding(0d, 0d, 10d, 0d);

        AnyChartView anyChartView = (AnyChartView) binding.chartView;
        anyChartView.setChart(chart);
    }

    private void createTestChart() {

        Log.i("CHART", "Creating Pie Chart");

        Pie pie = AnyChart.pie();

        List<DataEntry> data = new ArrayList<>();
        data.add(new ValueDataEntry("John", 10000));
        data.add(new ValueDataEntry("Jake", 12000));
        data.add(new ValueDataEntry("Peter", 18000));

        pie.data(data);

        AnyChartView anyChartView = (AnyChartView) binding.chartView;
        anyChartView.setChart(pie);

        Log.i("CHART", "Finished Chart");

    }
}

File: app/src/main/java/ght/app/datalogger/ChartFragment.kt
package ght.app.datalogger

import android.animation.ObjectAnimator
import android.content.Context
import android.graphics.Color.BLACK
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import com.anychart.AnyChart
import com.anychart.chart.common.dataentry.DataEntry
import com.anychart.chart.common.dataentry.ValueDataEntry
import com.anychart.charts.Cartesian
import com.google.android.material.color.MaterialColors
import com.google.android.material.snackbar.Snackbar
import ght.app.datalogger.data.logSystem.IntfGuiListener
import ght.app.datalogger.data.logSystem.PrintOnMonitor
import ght.app.datalogger.databinding.FragmentChartBinding

/**
 * This [Fragment] is for creating the chart view and display the data from the
 * chosen unit.
 */

class ChartFragment : Fragment(), IntfGuiListener {

    private var _binding: FragmentChartBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    val model: UnitViewModel by activityViewModels()
    private var thiscontext: Context? = null;
    lateinit var chart: Cartesian;

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        if (container != null) {
            thiscontext = container.getContext()
        };

        _binding = FragmentChartBinding.inflate(inflater, container, false)

        chart = AnyChart.line()
        chart.background().fill(BLACK.toString(), 100);
        binding.chartView.setChart(chart)

        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Resolve color from theme attr
        val primaryColor = MaterialColors.getColor(
            view, R.attr.colorSecondary
        )

        val model: UnitViewModel by activityViewModels()

        chart = drawChart(chart, model, view)

        if (model.getConnectionState()) {
            binding.fabRefresh.show()
            model.getActiveUnit()?.let {
                model.addListener(
                    this@ChartFragment,
                    IntfGuiListener.LogUnitEvent.CONNECTION_LOST,
                    it
                )
                model.addListener(
                    this@ChartFragment,
                    IntfGuiListener.LogUnitEvent.CMDFEEDBACK_RECEIVED,
                    it
                )
            }
        } else {
            binding.fabRefresh.hide()
        }
        binding.fabRefresh.setOnClickListener {
            thiscontext?.let { model.getActiveUnit()?.let { it1 -> model.sendCommand(2, it1, it) } }


            //ObjectAnimator.ofFloat(binding.fabRefresh, "rotation", 0f, 360f).start();
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()

        _binding = null
    }

    /**
     * Methode is updating the Chart with the received values of the LoggingUnit datas
     * @param chart that needs to be updated (should be the same chart as the one that is giving back to the at return-value)
     * @param model that contains the active Unit
     * @param view current view
     * @return the adress of the updated chart (this is not a new object, but the same as the one that has given into this Method as a param called chart)
     */
    private fun drawChart(chart: Cartesian, model: UnitViewModel, view: View) : Cartesian {
        //val chart = AnyChart.line()
        chart.title( model.getActiveUnit() + "  (" + (model.getTrendDataSize()-1) + " Datenpunkte)")
        chart.yAxis(0).title("Sensorwert")
        chart.xScroller(true);
        //chart.autoRedraw(true);

        // set first series data

        val data: MutableList<DataEntry> = getData()

        var line = chart.line(data);
        // set stroke thickness 5px and stroke color as darkOrange
        line.stroke("2 DarkOrange");

        if (data.isEmpty()) {
            data.add(ValueDataEntry("10-10-2020 10:10", 0))
            Snackbar.make(view, "Keine Daten vorhanden. " +
                    "Bitte zuerst verbinden und von Unit laden", Snackbar.LENGTH_LONG)
                .show()
        }

        chart.data(data)
        chart.draw(true)

        return chart
    }

    /**
     * Methode to get the Datas of the LoggingUnit-logDataList and writes into a MutableList
     * @return MutableList with all the Logdatas in it.
     */
    private fun getData() : MutableList<DataEntry> {
        val logData = model.getTrendData()

        val data: MutableList<DataEntry> = ArrayList()

        for (i in 1 until logData.size) {
            if (logData[i].contains("*")) {
                continue
            } else {
                data.add(toDataEntry(logData[i]))
            }
        }
        return data
    }

    private fun toDataEntry(data: String): ValueDataEntry {
        val x = data.substringBefore(";")
        val value = data.substringAfter(";").toFloat()
        return ValueDataEntry(x, value)
    }

    override fun loggingUnitEvent(
        lue: IntfGuiListener.LogUnitEvent?,
        value: Int,
        unitName: String?) {
        when (lue) {
            IntfGuiListener.LogUnitEvent.CONNECTION_LOST -> {
                binding.fabRefresh.hide()
                model.getActiveUnit()?.let {
                    model.removeListener(
                        this@ChartFragment,
                        IntfGuiListener.LogUnitEvent.CONNECTION_LOST,
                        it
                    )
                    model.removeListener(
                        this@ChartFragment,
                        IntfGuiListener.LogUnitEvent.CMDFEEDBACK_RECEIVED,
                        it
                    )
                }
            }
            IntfGuiListener.LogUnitEvent.CMDFEEDBACK_RECEIVED -> {
                if (value == 2) {
                    //Update Treandview if new Datas got received
                    view?.let { drawChart(chart, model, it) }
                    //ObjectAnimator.ofFloat(binding.fabRefresh, "rotation", 0f, 360f).end();
                }
            }
        }
    }

}